destination/destination.go

```diff
@@ -0,0 +1,148 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package destination
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-snowflake/destination/schema"
+	"github.com/conduitio-labs/conduit-connector-snowflake/destination/writer"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-errors/errors"
+)
+
+const (
+	defaultBatchDelay = time.Second * 5
+	defaultBatchSize  = 1000
+)
+
+type Destination struct {
+	sdk.UnimplementedDestination
+
+	Config Config
+	Writer writer.Writer
+	Schema schema.Schema
+}
+
+// NewDestination creates the Destination and wraps it in the default middleware.
+func NewDestination() sdk.Destination {
+	// This is needed to override the default batch size and delay defaults for this destination connector.
+	middlewares := sdk.DefaultDestinationMiddleware()
+	for i, m := range middlewares {
+		switch dest := m.(type) {
+		case sdk.DestinationWithBatch:
+			dest.DefaultBatchDelay = defaultBatchDelay
+			dest.DefaultBatchSize = defaultBatchSize
+			middlewares[i] = dest
+		default:
+		}
+	}
+
+	return sdk.DestinationWithMiddleware(&Destination{}, middlewares...)
+}
+
+func (d *Destination) Parameters() map[string]sdk.Parameter {
+	return Config{}.Parameters()
+}
+
+func (d *Destination) Configure(ctx context.Context, cfg map[string]string) error {
+	sdk.Logger(ctx).Debug().Msg("Configuring Destination Connector.")
+
+	// TODO: add configuration for ordering column (aka primary key)
+	// right now, we will automatically detect the key from the key within the record,
+	// but it would be great to have flexibility in case the user wants to key on a different
+
+	err := sdk.Util.ParseConfig(cfg, &d.Config)
+	if err != nil {
+		return errors.Errorf("failed to parse destination config: %w", err)
+	}
+
+	return nil
+}
+
+// Open prepares the plugin to receive data from given position by
+// initializing the database connection and creating the file stage if it does not exist.
+func (d *Destination) Open(ctx context.Context) error {
+	writer, err := writer.NewSnowflake(ctx, &writer.SnowflakeConfig{
+		Prefix:     d.Config.NamingPrefix,
+		PrimaryKey: d.Config.PrimaryKey,
+		Stage:      d.Config.Stage,
+		TableName:  d.Config.Table,
+		Connection: d.Config.Connection,
+		Format:     d.Config.Format,
+	})
+	if err != nil {
+		return errors.Errorf("failed to open connection to snowflake: %w", err)
+	}
+
+	d.Writer = writer
+
+	return nil
+}
+
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	if d.Schema == nil {
```

Comment:
The schema appears to be only initialized, but never actually used?

Reply:
@hariso correct, I didn't push the changes because then discovered `nil` type issue and decided to wait to see how we can fill this in.